### PR TITLE
Soaring loiter graceful state changes

### PR DESCRIPTION
**Pull request by magicrub**
Notify users why a soaring loiter exits happen, and allow better control over what happens in the different scenarios including exiting on heading alignment in some scenarios

### CODE REVIEW

**Comment by robustini** [comment]
Love it, thanks @magicrub!!!

**Comment by robustini** [comment]
@magicrub, your PR works perfectly in the X-Plane sim with SITL, unfortunately it is not easy to understand why the SOAR while seeing a huge loss of height during the Loiter does not always switch to FBWB, the first few times it does then it stops, continuing to make the glider lose altitude , I therefore believe that there is a problem in the SOAR library, I therefore believe that @samuelctabor must check the code.
Unfortunately the Planner continues to report problems with AHRS during SITL, so i don't see the messages of your PR, however I attach the binary log and a video showing the correct functioning.

https://youtu.be/zOJVnTfiBr8

[00000004.ZIP](https://github.com/ArduPilot/ardupilot/files/3260402/00000004.ZIP)

**Comment by samuelctabor** [comment]
Nice work. I'm travelling right now but will test in next could of days.

Marco  I've seen that before. One possible cause is fixed by https://github.com/ArduPilot/ardupilot/pull/11414

Could you see if the same problem can occur in that branch?

**Comment by robustini** [comment]
> 
> 
> Nice work. I'm travelling right now but will test in next couple of days.
> 
> Marco I've seen that before. One possible cause is fixed by #11414
> 
> Could you see if the same problem can occur in that branch?

I can prove it, but in this branch there aren't any of your last fixes.

**Comment by robustini** [comment]
I can test it, but in this branch there aren't any of your last fixes.

**Comment by magicrub** [comment]
I can add it, would you like that?

**Comment by robustini** [comment]
> 
> 
> I can add it, would you like that?

Yes Tom. thanks!

**Comment by magicrub** [comment]
@samuelctabor had two branches,here they are rebased on master
https://github.com/magicrub/ardupilot/tree/Soaring-coordinateChange-master
https://github.com/magicrub/ardupilot/tree/Soaring-varioImprovements-master

and attached binaries



[bin.zip](https://github.com/ArduPilot/ardupilot/files/3262870/bin.zip)

File: ArduPlane/soaring.cpp [hunk 1]
```diff
@@ -43,57 +43,124 @@ void Plane::update_soaring() {
...
-            g2.soaring_controller.get_target(next_WP_loc);
+        const SoaringController::LoiterStatus loiterStatus = g2.soaring_controller.check_cruise_criteria();
+
+        if (loiterStatus == SoaringController::LoiterStatus::THERMAL_GOOD_TO_KEEP_LOITERING ||
+            loiterStatus == SoaringController::LoiterStatus::SOARING_DISABLED) {
+            break;
         }
+
+        const char* strTooHigh = "Soaring: Too high, restoring ";
+        const char* strTooLow =  "Soaring: Too high, restoring ";
```
Comment (samuelctabor) [inline ArduPlane/soaring.cpp]: Typo, should be "Soaring: Too low, restoring "

**Comment by samuelctabor** [comment]
@magicrub Do you have a reason for the logic on when the heading has to be lined up?

Currently from FBWB:
 - too high only

From CRUISE:
 - never

From AUTO:
 - too weak only

Might simplify logic to always line up with the appropriate heading for the mode

**Comment by samuelctabor** [comment]
It's possible to get a lot of messages when the thermal estimate deteriorates as the inner soaring loop will print a message each time, until the heading is lined up for exit.

Screenshot from 2019-06-07 08-36-37

**Comment by robustini** [comment]
I'm testing now the "Soaring-varioImprovements-master" branch on the SITL with the addition of @magicrub, so far no problem.
Automatic switching between FBWB and Loiter takes place regularly, and airspeed is respected...

immagine

**Comment by robustini** [comment]
However, I continue to reiterate that the exit from the Loiter and the return in FBWB is too bland, ie it tries to stay in thermal even when the glider is losing a lot of altitude, and in that case it should go out and look elsewhere.
@samuelctabor, imho this is a problem in the soaring library.

**Comment by samuelctabor** [comment]
> However, I continue to reiterate that the exit from the Loiter and the return in FBWB is too bland, ie it tries to stay in thermal even when the glider is losing a lot of altitude, and in that case it should go out and look elsewhere.
> @samuelctabor, imho this is a problem in the soaring library.

Hi Marco,

Could you upload the log - in the previous logs the problem should have been fixed in the varioImprovents branch.

**Comment by robustini** [comment]
> 
> 
> > However, I continue to reiterate that the exit from the Loiter and the return in FBWB is too bland, ie it tries to stay in thermal even when the glider is losing a lot of altitude, and in that case it should go out and look elsewhere.
> > @samuelctabor, imho this is a problem in the soaring library.
> 
> Hi Marco,
> 
> Could you upload the log - in the previous logs the problem should have been fixed in the varioImprovents branch.

Here the log of my latest test with that branch (with Tom's addition), and "Too Low" manual fix.

https://1drv.ms/u/s!Ao6L05rilbXuhEVCga1rH_rdBh0s?e=bh9zK5

**Comment by robustini** [comment]
I've not looked at how the code inside the library reasons, but in my opinion when entering Loiter it should memorize the height, make its calculations and if exceeded the time declared in "SOAR_MIN_THML_S" and and the height is lower than when he entered Loiter it should exit immediately (for example return in FBWB) and search elsewhere.
If it is already so it does not.

**Comment by samuelctabor** [comment]
Currently it looks only at the estimated thermal parameters and calculates the corresponding climb rate potential at the loitering radius. Sometimes if there is poor updraft measured the estimate will think it is a strong but very narrow thermal, as this also fits the measurements well.

There is a bug in master where the loitering radius is saved as zero, that means a strong very narrow thermal is calculated as having good climb potential and the thermal is not abandoned.

A total altitude gained check as you suggest would be a nice backup watchdog.

**Comment by robustini** [comment]
Well, then if you want me to be able to perform accurate tests fix that bug and consider what I wrote above.
If you create a new branch with all this changes and the related fix, maybe today i've time to test it in reality, after some tests on SITL.

**Comment by samuelctabor** [comment]
Agreed, I'll create a branch with all soaring related fixes.

**Comment by samuelctabor** [comment]
Hi marco,

Could you upload those logs directly here, having trouble downloading from OneDrive.

**Comment by robustini** [comment]
immagine

**Comment by robustini** [comment]
Add me in Skype, search my name and surname.

**Comment by magicrub** [comment]
>Add me in Skype, search my name and surname.

Add me too! Let's make a group chat

File: ArduPlane/mode_loiter.cpp [hunk 2]
```diff
@@ -25,3 +26,46 @@ void ModeLoiter::update()
...
+      waypoint. We use 20 degrees (+/-10 deg) of margin so that
+      we can handle 200 degrees/second of yaw.
+
+      After every full circle, extend acceptance criteria to ensure
+      aircraft will not loop forever in case high winds are forcing
+      it beyond 200 deg/sec when passing the desired exit course
+    */
+
+    // Use integer division to get discrete steps
+    const int32_t expanded_acceptance = 1000 * (plane.loiter.sum_cd / 36000);
```
Comment (samuelctabor) [inline ArduPlane/mode_loiter.cpp]: plane.loiter.sum_cd can be negative

Comment (magicrub) [reply ArduPlane/mode_loiter.cpp]: some random guy from the internet fixed it!
https://github.com/magicrub/ardupilot/commit/1fcfc9a35cd1e98fc9694b241d25a4f4772f19b6

**Comment by samuelctabor** [comment]
Haha. Fixes the negative loiter.sum_cd, not the loiter.sum_cd counting from loiter start, rather than exit condition becoming true

**Comment by magicrub** [comment]
Closing. This has moved to a new branch which is being tested.

**Pull request closed without merging**